File: tba/src/main/java/de/gaalop/productComputer/bladeProdComputing/GeoProductCalculator.java
package de.gaalop.productComputer.bladeProdComputing;

import de.gaalop.dfg.Expression;
import de.gaalop.dfg.Negation;
import de.gaalop.dfg.Variable;
import de.gaalop.productComputer.AlgebraDefinitionTC;
import de.gaalop.productComputer.bladeOperations.BladeNormaliser;
import de.gaalop.productComputer.dataStruct.TCBlade;
import de.gaalop.tba.Algebra;
import java.util.Arrays;
import java.util.LinkedList;

/**
 * Calculates the geometric product of two blades
 * @author Christian Steinmetz
 */
public class GeoProductCalculator extends ProductCalculator {

    public GeoProductCalculator(Algebra algebra, AlgebraDefinitionTC algebraDefinition) {
        super(algebra, algebraDefinition);
    }

    @Override
    public Expression calculateProductBlades(TCBlade blade1, TCBlade blade2) {
        //approach method from the dissertation: Daniel Fontijne, "Efficient Implementation of Geometric Algebra" (2007)

        LinkedList<String> list1 = new LinkedList<String>(Arrays.asList(blade1.getBase()));
        LinkedList<String> list2 = new LinkedList<String>(Arrays.asList(blade2.getBase()));

        boolean negate = false;

        list1.remove("1");
        list2.remove("1");

        String element;
        while ((element = commonElement(list1, list2)) != null) {
            int disp1 = list1.size()-list1.indexOf(element)-1;
            int disp2 = list2.indexOf(element);
            if (((disp1+disp2) % 2) == 1)
                negate = !negate;
            list1.remove(element);
            list2.remove(element);
            if (algebraDefinition.baseSquares.get(element) == -1) negate = !negate;
        }

        LinkedList<String> merged = new LinkedList<String>();
        merged.addAll(list1);
        merged.addAll(list2);
        String[] arr = merged.toArray(new String[0]);
        
        BladeNormaliser normaliser = new BladeNormaliser(algebraDefinition);
        byte normalizeFactor = normaliser.normalize(arr);
        if (negate) normalizeFactor *= -1;
        int index = getIndexFromBase(arr);

        Variable variable = new Variable("G"+index);
        return (normalizeFactor == 1) ? variable : new Negation(variable);
    }

    /**
     * Returns a common element of two lists
     * @param list1 The first list
     * @param list2 The second list
     * @return A common element
     */
    private String commonElement(LinkedList<String> list1, LinkedList<String> list2) {
        for (String cur: list1)
            if (list2.contains(cur))
                return cur;
        return null;
    }


}

File: productComputerC/GeoProductCalculator.cpp
/*
 * GeoProductCaluclator.cpp
 *
 *  Created on: 22.11.2011
 *      Author: christian
 */

#include "GeoProductCalculator.h"

GeoProductCalculator::GeoProductCalculator() {
}

GeoProductCalculator::~GeoProductCalculator() {
}

void GeoProductCalculator::calculate(const Blade& blade1, const Blade& blade2, unordered_map<int,int>& baseSquares, SumOfBlades& result) {
	//approach method from the dissertation: Daniel Fontijne, "Efficient Implementation of Geometric Algebra" (2007)

	intList list1;
	intList list2;
	VectorMethods::intVectorToList(blade1.baseVectors,list1);
	VectorMethods::intVectorToList(blade2.baseVectors,list2);

	bool negate = false;

	list1.remove(0);
	list2.remove(0);


	int element;
	int i1;
	int i2;
	while ((element = commonElement(list1, list2, i1, i2)) != -1) {
		int disp1 = list1.size()-i1-1;
		int disp2 = i2;
		if (((disp1+disp2) % 2) == 1)
			negate = !negate;
		list1.remove(element);
		list2.remove(element);
		if (baseSquares[element] == -1) negate = !negate;
	}

	Blade b;
	intVector& merged = b.baseVectors;
	merged.reserve(list1.size()+list2.size());
	for (intList::const_iterator ci = list1.begin(); ci != list1.end(); ++ci)
		merged.push_back(*ci);
	for (intList::const_iterator ci = list2.begin(); ci != list2.end(); ++ci)
		merged.push_back(*ci);

	float prefactor = blade1.getPrefactor()*blade2.getPrefactor();
	if (negate) prefactor *= -1;
	b.setPrefactor(prefactor);

	result.addBlade(b);
}


/**
 * Returns a common element of two lists
 * @param list1 The first list
 * @param list2 The second list
 * @return A common element
 */
int GeoProductCalculator::commonElement(const intList& list1, const intList& list2, int& i1, int& i2) {
	i1 = 0;
	i2 = 0;
	for (intList::const_iterator ci1 = list1.begin(); ci1 != list1.end(); ++ci1) {
		i2=0;
		for (intList::const_iterator ci2 = list2.begin(); ci2 != list2.end(); ++ci2) {
			if (*ci1 == *ci2)
				return *ci1;
			i2++;
		}
		i1++;
	}
	return -1;
}
